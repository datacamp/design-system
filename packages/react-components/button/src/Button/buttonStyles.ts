import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css, SerializedStyles } from '@emotion/core';
import tinycolor from 'tinycolor2';

interface IntentMap {
  cta: string;
  danger: string;
  neutral: string;
  success: string;
  warning: string;
}

type Intent = keyof IntentMap;

// CUSTOM TOKENS
const baseColors: IntentMap = {
  cta: tokens.color.primary.green.value.hex,
  danger: tokens.color.primary.red.value.hex,
  neutral: tokens.color.primary.navy.value.hex,
  success: tokens.color.primary.green.value.hex,
  warning: tokens.color.primary.orange.value.hex,
};

const hoverColors: IntentMap = {
  cta: '#E7B743',
  danger: '#B64242',
  neutral: '#2E8FAB',
  success: '#29A360',
  warning: '#C87400',
};

const getFocusColor = (intent: Intent): string =>
  tinycolor(baseColors[intent])
    .setAlpha(0.2)
    .toRgbString();

// BASE STYLES
const baseStyle = css({
  ':active': { transform: 'perspective(1px) scale(0.975)' },
  ':disabled, :hover:disabled, :active:disabled': {
    transform: 'none',
  },
  ':focus': { outline: '0' },
  alignItems: 'center',
  border: 'none',
  borderRadius: '4px',
  borderStyle: 'solid',
  borderWidth: '1px',
  cursor: 'pointer',
  display: 'inline-flex',
  justifyContent: 'center',
  textDecoration: 'none',
  textTransform: 'capitalize',
  transition: '0.15s',
  verticalAlign: 'baseline',
  whiteSpace: 'nowrap',
});

const baseLoadingStyle = css({
  ':active': { transform: 'none' },
  ':focus, :hover:focus, :hover': {
    cursor: 'wait',
  },
  position: 'relative',
});

const getSize = (size: 'small' | 'medium' | 'large'): SerializedStyles => {
  if (size === 'large') {
    return css({ padding: '0 31px' });
  }
  return css({ padding: '0 15px' });
};

const getIconSize = (size: 'small' | 'medium' | 'large'): SerializedStyles => {
  if (size === 'large') {
    return css({ padding: '19px' });
  }
  if (size === 'small') {
    return css({ padding: '8px' });
  }
  return css({ padding: '14px' });
};

const fontSizes = {
  large: { fontSize: '20px', lineHeight: '62px' },
  medium: { fontSize: '16px', lineHeight: '46px' },
  small: { fontSize: '16px', lineHeight: '34px' },
};

const getDefaultStyle = (intent: Intent, enabled: boolean): SerializedStyles =>
  css(
    {
      ':focus': {
        boxShadow: `0 0 0 4px ${getFocusColor(intent)}}`,
      },
      backgroundColor: 'transparent',
      borderColor: baseColors[intent],
      color: baseColors[intent],
    },
    enabled && {
      ':hover': {
        borderColor: hoverColors[intent],
        color: hoverColors[intent],
      },
    },
  );

const getPrimaryStyle = (intent: Intent, enabled: boolean): SerializedStyles =>
  css(
    getDefaultStyle(intent, enabled),
    {
      backgroundColor: baseColors[intent],
    },
    enabled && {
      ':hover': {
        backgroundColor: hoverColors[intent],
        borderColor: hoverColors[intent],
      },
    },
  );

const getAppearanceStyle = (
  appearance: 'primary' | 'default',
  intent: Intent,
  enabled: boolean,
): SerializedStyles =>
  appearance === 'primary'
    ? getPrimaryStyle(intent, enabled)
    : getDefaultStyle(intent, enabled);

const getDisabledStyle = (
  appearance: 'primary' | 'default',
  intent: Intent,
): SerializedStyles =>
  css(
    { borderColor: baseColors[intent], cursor: 'not-allowed', opacity: 0.2 },
    appearance === 'primary' && {
      backgroundColor: baseColors[intent],
      borderColor: baseColors[intent],
    },
  );

export {
  baseColors,
  baseLoadingStyle,
  baseStyle,
  fontSizes,
  getAppearanceStyle,
  getDisabledStyle,
  getIconSize,
  getSize,
  hoverColors,
};
