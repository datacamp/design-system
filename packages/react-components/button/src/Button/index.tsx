import * as Icons from '@datacamp/waffles-icons';
import { Text } from '@datacamp/waffles-text';
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { computeDataAttributes } from '@datacamp/waffles-utils';
import { css, SerializedStyles } from '@emotion/core';
import { childrenOfType, nChildren } from 'airbnb-prop-types';
import PropTypes from 'prop-types';
import React, { ComponentProps, ReactElement, useRef, useState } from 'react';
import { useUIDSeed } from 'react-uid';
import { mergeRefs } from 'use-callback-ref';

import Tooltip from '../Tooltip';
import {
  baseColors,
  baseLoadingStyle,
  baseStyle,
  fontSizes,
  getAppearanceStyle,
  getDisabledStyle,
  getIconSize,
  getSize,
} from './buttonStyles';
import Spinner from './spinner';

type IconType = typeof Icons.AddCircleIcon; // Could use any Icon here
type IconElement = ReactElement<ComponentProps<IconType>, IconType>;

interface BaseButtonProps {
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
  /**
   * As with all the other waffles components, dataAttributes can be used to set
   * data- html attributes on the element.
   */
  dataAttributes?: { [key: string]: string };
  /**
   * The button can be disabled by passing "disabled" as a prop.
   */
  disabled?: boolean;
  /**
   * By passing "loading" as a prop, the button will be disabled and it will
   * render a spinner.
   */
  loading?: boolean;
  /**
   * The size of the button to render.
   */
  size?: 'small' | 'medium' | 'large';
  /**
   * When tooltipText is provided, this sets the colour of the tooltip. It
   * should be used make the tooltip visible on different colour backgrounds.
   */
  tooltipAppearance?: 'light' | 'dark';
  /**
   * When tooltipText is provided, this determines the position of that text.
   * "top", "bottom", "left", and "right" will position the tooltip at the
   * centre of that side. The other styles can be used to prevent the tooltip
   * from overflowing when it is longer than the button.
   */
  tooltipPosition?:
    | 'bottom'
    | 'top'
    | 'left'
    | 'right'
    | 'bottomRight'
    | 'bottomLeft'
    | 'topLeft'
    | 'topRight';
  /**
   * An additional description to show on hover or focus.
   */
  tooltipText?: string;
}

interface LinkButtonProps {
  href: string;
  target?: '_blank' | undefined;
  type: 'link';
}

interface ButtonButtonProps {
  onClick: () => void;
  type?: 'button' | undefined;
}

interface SubmitButtonProps {
  type: 'submit';
}

interface StringChildProps {
  ariaLabel?: string;
  children: string;
}

interface IconChildProps {
  ariaLabel: string;
  /**
   * When providing a react element as a child, it can only be one of the Icon
   * components exposed by @datacamp/waffles-icons
   */
  children: IconElement;
}

interface DefaultProps {
  appearance?: 'default';
  intent?: intents;
}

interface PrimaryProps {
  appearance: 'primary';
  intent?: intents | 'cta';
}

type intents = 'neutral' | 'danger' | 'success' | 'warning';

interface IconTextChildProps {
  ariaLabel?: string;
  children: [string, IconElement] | [IconElement, string];
}

export type ButtonProps = BaseButtonProps &
  (ButtonButtonProps | LinkButtonProps | SubmitButtonProps) &
  (StringChildProps | IconChildProps | IconTextChildProps) &
  (DefaultProps | PrimaryProps);

const InternalButton = (
  props: ButtonProps & { innerRef?: React.Ref<any> },
): React.ReactElement => {
  const {
    appearance = 'default',
    ariaLabel,
    children,
    className,
    dataAttributes,
    disabled = false,
    innerRef,
    intent = 'neutral',
    loading = false,
    size = 'medium',
    tooltipAppearance = 'dark',
    tooltipText,
    tooltipPosition = 'bottom',
  } = props;

  const buttonRef = useRef<HTMLElement>();
  const [hasFocus, setHasFocus] = useState(false);
  const [hasHover, setHasHover] = useState(false);
  const uidSeed = useUIDSeed();
  const tooltipId = uidSeed('button-tooltip');
  const outlineTextColor = tokens.color.opaque.greyDark.value.hex;
  const ctaTextColor = intent === 'cta' ? outlineTextColor : 'white';
  const outlineIconColor = disabled ? baseColors[intent] : 'currentColor';
  const textColor = appearance === 'primary' ? ctaTextColor : outlineTextColor;
  const iconColor = appearance === 'primary' ? ctaTextColor : outlineIconColor;

  const baseTextStyle = css(fontSizes[size], {
    color: loading ? 'transparent' : textColor,
    fontWeight: 'bold',
  });
  const margin =
    size === 'large' ? tokens.size.space[16].value : tokens.size.space[8].value;
  const getTextStyleWithMargin = (i: number | undefined): SerializedStyles => {
    if (i === 1) {
      return css(baseTextStyle, {
        marginLeft: margin,
      });
    }
    return css(baseTextStyle, {
      marginRight: margin,
    });
  };
  const parsedDataAttributes = computeDataAttributes(dataAttributes);
  const buttonStyle = css(
    baseStyle,
    getAppearanceStyle(appearance, intent, !loading && !disabled),
    disabled && getDisabledStyle(appearance, intent),
    loading && baseLoadingStyle,
    React.Children.count(children) > 1 || typeof children === 'string'
      ? getSize(size)
      : getIconSize(size),
  );

  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
  const { Element, ...otherProps } = (() => {
    switch (props.type) {
      case 'link':
        return {
          Element: 'a',
          href: props.href,
          target: props.target,
        } as const;
      case 'submit':
        return { Element: 'button', type: 'submit' } as const;
      case 'button':
      default:
        return {
          Element: 'button',
          onClick: !loading ? props.onClick : undefined,
          type: 'button',
        } as const;
    }
  })();

  const tooltipVisible = !(disabled || loading) && (hasFocus || hasHover);

  return (
    <>
      <Element
        ref={innerRef ? mergeRefs([buttonRef, innerRef]) : buttonRef}
        aria-describedby={tooltipText && tooltipVisible ? tooltipId : undefined}
        aria-label={ariaLabel}
        className={className}
        css={buttonStyle}
        disabled={disabled || loading}
        onBlur={() => setHasFocus(false)}
        onFocus={() => setHasFocus(true)}
        onMouseEnter={() => setHasHover(true)}
        onMouseLeave={() => setHasHover(false)}
        {...otherProps}
        {...parsedDataAttributes}
      >
        {loading && (
          <Spinner
            css={{ position: 'absolute' }}
            inverted={appearance === 'primary' && intent !== 'cta'}
          />
        )}
        {React.Children.map(children, (child, i) =>
          typeof child === 'string' ? (
            <Text
              css={
                React.Children.count(children) > 1
                  ? getTextStyleWithMargin(i)
                  : baseTextStyle
              }
            >
              {child}
            </Text>
          ) : (
            React.cloneElement(child, {
              'aria-hidden': true, // hide icon from screen reader so only ariaLabel or button text is read.
              color: loading ? 'transparent' : child.props.color ?? iconColor,
              size: size === 'large' ? 24 : 18,
              title: '', // remove tooltip from icon within button
            })
          ),
        )}
      </Element>
      {tooltipText && (
        <Tooltip
          appearance={tooltipAppearance}
          id={tooltipId}
          position={tooltipPosition}
          target={buttonRef}
          visible={tooltipVisible}
        >
          {tooltipText}
        </Tooltip>
      )}
    </>
  );
};

// additional prop-types validation that the child is either icon or string
const iconValidator = childrenOfType(...Object.values(Icons));

InternalButton.propTypes = {
  /**
   * When the appearance is "primary", the button will have a filled colour.
   */
  appearance: PropTypes.oneOf(['default', 'primary']),
  /**
   * Set the aria-label on the rendered element. This is required when using a single icon as a child.
   */
  ariaLabel: PropTypes.string,
  /**
   * The content of the button. This can either be text, and icon, or a
   * combination of the two. When this is one of the components exposed by
   * @datacamp/waffles-icons, the button will render as a square. In this
   * situation ariaLabel is required.
   */
  children: PropTypes.oneOfType([
    nChildren(
      2,
      PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, iconValidator])),
    ),
    PropTypes.string,
    iconValidator,
  ]).isRequired,
  /**
   * The destination of the link. Only available when type="link".
   */
  href: PropTypes.string,
  /**
   * Defines the color of the button. The cta intent can be used only with
   * the primary appearance.
   */
  intent: PropTypes.oneOf(['neutral', 'danger', 'success', 'warning', 'cta']),
  /**
   * The callback fired when the button is clicked. Only available when
   * type="button".
   */
  onClick: PropTypes.func,
  /**
   * Specifies where to open the linked document. Only available when
   * type="link".
   */
  target: PropTypes.string,
  /**
   * Determines what is rendered, and what other props are available. The type
   * "submit" should be used for form submissions. The type "link" will render
   * an <a> tag, and should be used with an 'href'. And the type "button" should
   * be used with the 'onClick' prop.
   */
  type: PropTypes.oneOf(['button', 'link', 'submit']),
};

InternalButton.defaultProps = {
  appearance: 'default',
  disabled: false,
  intent: 'neutral',
  loading: false,
  size: 'medium',
  tooltipAppearance: 'dark',
  tooltipPosition: 'bottom',
  type: 'button',
};

const Button = React.forwardRef<any, ButtonProps>((props, ref) => (
  <InternalButton innerRef={ref} {...props} />
));

export default Button;
