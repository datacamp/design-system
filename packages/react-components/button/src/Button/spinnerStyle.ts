import { css, keyframes } from '@emotion/core';

const spinner = keyframes`0%
{ opacity: 0; }
100%
{ opacity: 1; }`;

const getNthChildSelector = (n: number): string =>
  `g:nth-child(${n}):not(style) /* emotion-disable-server-rendering-unsafe-selector-warning-please-do-not-use-this-the-warning-exists-for-a-reason */`;

const spinnerStyleObject: {
  [key: string]: { animation?: string; animationDelay?: string };
} = {
  g: {
    animation: `${spinner} 1s infinite`,
  },
};

spinnerStyleObject[getNthChildSelector(1)] = {
  animationDelay: '-0.9166666666666666s',
};
spinnerStyleObject[getNthChildSelector(2)] = {
  animationDelay: '-0.8333333333333334s',
};
spinnerStyleObject[getNthChildSelector(3)] = {
  animationDelay: ' -0.75s',
};
spinnerStyleObject[getNthChildSelector(4)] = {
  animationDelay: '-0.6666666666666666s',
};
spinnerStyleObject[getNthChildSelector(5)] = {
  animationDelay: '-0.5833333333333334s',
};
spinnerStyleObject[getNthChildSelector(6)] = {
  animationDelay: '-0.5s',
};
spinnerStyleObject[getNthChildSelector(7)] = {
  animationDelay: '-0.4166666666666667s',
};
spinnerStyleObject[getNthChildSelector(8)] = {
  animationDelay: '-0.3333333333333333s',
};
spinnerStyleObject[getNthChildSelector(9)] = {
  animationDelay: '-0.25s',
};
spinnerStyleObject[getNthChildSelector(10)] = {
  animationDelay: '-0.16666666666666666s',
};
spinnerStyleObject[getNthChildSelector(11)] = {
  animationDelay: '-0.08333333333333333s',
};
spinnerStyleObject[getNthChildSelector(12)] = {
  animationDelay: '0s',
};

const spinnerStyle = css(spinnerStyleObject);

export default spinnerStyle;
