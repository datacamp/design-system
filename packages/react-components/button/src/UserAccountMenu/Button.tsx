import { ChevronDownIcon, ChevronUpIcon } from '@datacamp/waffles-icons';
import tokens from '@datacamp/waffles-tokens';
import { css } from '@emotion/react';
import { useFocusRing } from '@react-aria/focus';
import { forwardRef } from 'react';

import AlertDot from './AlertDot';
import Avatar from './Avatar';
import { FULL_MENU_MEDIA_QUERY } from './constants';

const buttonStyle = css({
  alignItems: 'center',
  background: 'none',
  border: 0,
  borderRadius: 18,
  boxSizing: 'border-box',
  color: tokens.colors.navyText,
  cursor: 'pointer',
  display: 'inline-flex',
  [FULL_MENU_MEDIA_QUERY]: {
    ':hover': {
      backgroundColor: tokens.colors.beige200,
    },
    borderRadius: 15,
    height: 30,
    justifyContent: 'space-evenly',
    minWidth: 56,
    width: 56,
  },
  height: 36,
  minWidth: 36,
  outline: 0,
  padding: 0,
  width: 36,
});

const chevronStyle = css({
  display: 'none',
  [FULL_MENU_MEDIA_QUERY]: {
    display: 'block',
  },
});

type ButtonProps = {
  avatarUrl?: string;
  isOpen: boolean;
  showAlertDot?: boolean;
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ avatarUrl, isOpen, showAlertDot = false, ...restProps }, ref) => {
    const { focusProps, isFocusVisible } = useFocusRing();

    return (
      <button
        css={css(
          buttonStyle,
          isFocusVisible && {
            backgroundColor: tokens.colors.beige200,
            boxShadow: `0 0 0 2px ${tokens.colors.blueDark}`,
          },
        )}
        {...restProps}
        {...focusProps}
        ref={ref}
      >
        <Avatar avatarUrl={avatarUrl} />
        {isOpen ? (
          <ChevronUpIcon aria-hidden={true} css={chevronStyle} />
        ) : (
          <ChevronDownIcon aria-hidden={true} css={chevronStyle} />
        )}
        {showAlertDot && <AlertDot />}
      </button>
    );
  },
);

export default Button;
