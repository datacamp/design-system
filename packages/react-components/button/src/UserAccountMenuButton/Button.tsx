import { ChevronDownIcon, ChevronUpIcon } from '@datacamp/waffles-icons';
import tokens from '@datacamp/waffles-tokens';
import { css } from '@emotion/react';
import { useFocusRing } from '@react-aria/focus';
import { ButtonHTMLAttributes, forwardRef } from 'react';

import Avatar from './Avatar';

const buttonStyle = css({
  '@media (min-width: 821px)': {
    ':hover': {
      backgroundColor: tokens.colors.beige200,
    },
    borderRadius: 15,
    height: 30,
    justifyContent: 'space-evenly',
    minWidth: 56,
    width: 56,
  },
  alignItems: 'center',
  background: 'none',
  border: 0,
  borderRadius: 18,
  color: tokens.colors.navyText,
  cursor: 'pointer',
  display: 'inline-flex',
  height: 36,
  minWidth: 36,
  outline: 0,
  padding: 0,
  width: 36,
});

const chevronStyle = css({
  '@media (min-width: 821px)': {
    display: 'block',
  },
  display: 'none',
});

type ButtonProps = {
  isOpen: boolean;
} & ButtonHTMLAttributes<HTMLButtonElement>;

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  ({ isOpen, ...restProps }, ref) => {
    const { focusProps, isFocusVisible } = useFocusRing();

    return (
      <button
        css={css(
          buttonStyle,
          isFocusVisible && {
            backgroundColor: tokens.colors.beige200,
            boxShadow: `0 0 0 2px ${tokens.colors.blueDark}`,
          },
        )}
        {...restProps}
        {...focusProps}
        ref={ref}
      >
        <Avatar />
        {isOpen ? (
          <ChevronUpIcon aria-hidden={true} css={chevronStyle} />
        ) : (
          <ChevronDownIcon aria-hidden={true} css={chevronStyle} />
        )}
      </button>
    );
  },
);

export default Button;
