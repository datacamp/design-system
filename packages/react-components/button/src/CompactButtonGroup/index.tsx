import { ClassNames } from '@emotion/core';
import { childrenOfType } from 'airbnb-prop-types';
import PropTypes from 'prop-types';
import React from 'react';

import Button, { ButtonProps } from '../Button';

interface CompactButtonGroupProps {
  /**
   * The content of the CompactButtonGroup to render. This must be a set of
   * Button components. All Buttons within a CompactButtonGroup must have the
   * same appearance, intent, size & disabled state.
   */
  children: React.ReactElement<ButtonProps>[];
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
}

const CompactButtonGroup = ({
  children,
  className,
}: CompactButtonGroupProps): React.ReactElement => {
  const childrenProps = React.Children.map(children, ({ props }) => props);

  if (!childrenProps.every(({ size }) => size === childrenProps[0].size)) {
    throw Error('All Buttons in CompactButtonGroup must be the same size');
  }

  if (
    !childrenProps.every(
      ({ appearance, intent, disabled, size }) =>
        appearance === childrenProps[0].appearance &&
        disabled === childrenProps[0].disabled &&
        intent === childrenProps[0].intent &&
        size === childrenProps[0].size,
    )
  ) {
    throw Error(
      'All Buttons in compact CompactButtonGroup must have the same appearance, intent, size & disabled state',
    );
  }

  const buttonCount = React.Children.count(children);
  const radii = { 0: '4px 0px 0px 4px', [buttonCount - 1]: '0px 4px 4px 0px' };
  const isPrimaryButtonGroup = childrenProps[0].appearance === 'primary';
  return (
    <div
      className={className}
      css={{ display: 'inline-flex', whiteSpace: 'nowrap' }}
    >
      <ClassNames>
        {({ css }) =>
          React.Children.map(children, (child, index) => {
            return React.cloneElement(child, {
              className: css({
                ':active': { transform: 'none' },
                ':focus': { boxShadow: 'none' },
                borderLeft:
                  index > 0 && isPrimaryButtonGroup
                    ? '1px solid rgba(5, 25, 45, 0.3)'
                    : undefined,
                borderRadius: radii[index] || '0px', // only round corners on first & last child
                margin: 0,
                marginLeft: index > 0 ? '-2px' : undefined, // so only 1 border shows between buttons
              }),
            });
          })
        }
      </ClassNames>
    </div>
  );
};

CompactButtonGroup.propTypes = {
  children: PropTypes.arrayOf(childrenOfType(Button).isRequired).isRequired,
} as React.WeakValidationMap<CompactButtonGroupProps>;

export default CompactButtonGroup;
