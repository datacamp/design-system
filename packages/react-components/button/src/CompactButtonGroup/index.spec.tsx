import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import React from 'react';

import Button from '../Button';
import CompactButtonGroup from '.';

describe('CompactButtonGroup', () => {
  it('renders the buttons next to each other', async () => {
    const { container, getByText } = await axeRender(
      <CompactButtonGroup>
        <Button type="submit">Button 1</Button>
        <Button type="submit">Button 2</Button>
        <Button type="submit">Button 3</Button>
      </CompactButtonGroup>
    );

    const button1 = (getByText('Button 1') as HTMLElement).closest('button');
    const button2 = (getByText('Button 2') as HTMLElement).closest('button');
    const button3 = (getByText('Button 3') as HTMLElement).closest('button');

    expect(button1).toHaveStyle('margin-left: 0;');
    expect(button1).toHaveStyle('border-radius: 4px 0px 0px 4px;');

    expect(button2).toHaveStyle(`margin-left: -1px;`);
    expect(button2).toHaveStyle('border-radius: 0px;');

    expect(button3).toHaveStyle(`margin-left: -1px;`);
    expect(button3).toHaveStyle('border-radius: 0px 4px 4px 0px;');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('adds the className to the group', async () => {
    const testClass = 'test';
    const { container } = await axeRender(
      <CompactButtonGroup className={testClass}>
        <Button type="submit">Button 1</Button>
        <Button type="submit">Button 2</Button>
        <Button type="submit">Button 3</Button>
      </CompactButtonGroup>
    );

    expect(container.firstChild).toHaveClass(testClass);
  });
});
