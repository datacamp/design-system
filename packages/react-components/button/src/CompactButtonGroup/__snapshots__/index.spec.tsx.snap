// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompactButtonGroup renders the buttons next to each other 1`] = `
.emotion-6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  white-space: nowrap;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 2px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #05192d;
  color: #05192d;
  padding: 0 15px;
  border-radius: 4px 0px 0px 4px;
  margin: 0;
}

.emotion-1:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-1:disabled,
.emotion-1:hover:disabled,
.emotion-1:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-1:focus {
  outline: 0;
}

.emotion-1:hover {
  background-color: rgba(5,25,45,0.15);
  border-color: #05192d;
  color: #05192d;
}

.emotion-1:active {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-1:focus {
  box-shadow: none;
}

.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(5,25,45);
  font-family: Arial;
  font-family: Studio-Feixen-Sans-Variable;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  font-size: 16px;
  line-height: 44px;
  color: #05192d;
  font-weight: bold;
}

.emotion-3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 2px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #05192d;
  color: #05192d;
  padding: 0 15px;
  border-radius: 0px;
  margin: 0;
  margin-left: -2px;
}

.emotion-3:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-3:disabled,
.emotion-3:hover:disabled,
.emotion-3:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-3:focus {
  outline: 0;
}

.emotion-3:hover {
  background-color: rgba(5,25,45,0.15);
  border-color: #05192d;
  color: #05192d;
}

.emotion-3:active {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-3:focus {
  box-shadow: none;
}

.emotion-5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  border-radius: 4px;
  border-style: solid;
  border-width: 2px;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  -webkit-transition: 0.15s;
  transition: 0.15s;
  vertical-align: baseline;
  white-space: nowrap;
  background-color: transparent;
  border-color: #05192d;
  color: #05192d;
  padding: 0 15px;
  border-radius: 0px 4px 4px 0px;
  margin: 0;
  margin-left: -2px;
}

.emotion-5:active {
  -webkit-transform: perspective(1px) scale(0.975);
  -ms-transform: perspective(1px) scale(0.975);
  transform: perspective(1px) scale(0.975);
}

.emotion-5:disabled,
.emotion-5:hover:disabled,
.emotion-5:active:disabled {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-5:focus {
  outline: 0;
}

.emotion-5:hover {
  background-color: rgba(5,25,45,0.15);
  border-color: #05192d;
  color: #05192d;
}

.emotion-5:active {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.emotion-5:focus {
  box-shadow: none;
}

<div
  class="emotion-6"
>
  <button
    class="emotion-1"
    type="submit"
  >
    <span
      class="emotion-0"
    >
      Button 1
    </span>
  </button>
  <button
    class="emotion-3"
    type="submit"
  >
    <span
      class="emotion-0"
    >
      Button 2
    </span>
  </button>
  <button
    class="emotion-5"
    type="submit"
  >
    <span
      class="emotion-0"
    >
      Button 3
    </span>
  </button>
</div>
`;
